File: 2019/question_2.cpp
/* Score: 15/15 */

#include <iostream>
#include <map>

using namespace std;

map<int, int> primes;

bool isPrime(int num) { //checks if number is prime

    if(primes[num] == 1){ //if prime
        return true;
    } else if(primes[num] == 2) { //if not prime
        return false;
    } else { //if we do not know its status
        bool prime = true;

        int divisor = 2;
        if(num%divisor == 0) { //if even
            prime = false;
        } else { //if odd
            divisor = 3;
            while(prime && divisor <= num/2) {
                if(num%divisor == 0) {
                    prime = false;
                }
                divisor += 2;
            }
        }

        //stores result in map for quicker access in the future
        if(prime) {
            primes[num] = 1; //true
        } else {
            primes[num] = 2; //false
        }

        return prime;
    }
}

int main() {
    //gets inputs
    int inputs = 0;
    cin >> inputs;

    int * arr = new int[inputs];

    for(int i = 0; i < inputs; i++) {
        cin >> arr[i];
    }

    int doubled = 0;
    bool found;
    int lowerBound = 0, upperBound = 0;

    for(int i = 0; i < inputs; i++) { //goes through each number in the array
        doubled = arr[i]*2;
        lowerBound = 2; //lower bound of the average
        upperBound = doubled-lowerBound; //upper bound of the average
        found = false;

        while(!found && lowerBound <= arr[i]) { //goes through possible value pairs until a value is found where both bounds are prime
            if(isPrime(lowerBound) && isPrime(upperBound)) { //if both bounds are prime
                cout << lowerBound << " " << upperBound << endl;
                found = true;
            }

            if(lowerBound == 2) {
                upperBound--;
                lowerBound++;
            } else {
                upperBound -= 2;
                lowerBound += 2;
            }
        }
    }
}

File: 2019/question_2.py
# Score: 6/15

# dictionary containing all numbers and their status as prime (true) or not prime (false)
primes = {}

def isPrime(num): # checks if the number is a prime number
    try: # checks dictionary to see if number already exists
        if primes[num]: # if prime
            return True
        else: # if not prime
            return False
    except KeyError: # if number not in dictionary
        prime = True
        divisor = 2

        if num%divisor == 0 and num != 2: # checks if even
            prime = False

        divisor = 3
        # checks all possible odd divisors
        while prime and divisor <= num/2 and num != 2:
            if num%divisor == 0:
                prime = False

            divisor = divisor+2

        # sets number in dictionary
        if prime:
            primes[num] = True
        else:
            primes[num] = False

        return prime

# gets input
input_number = int(input())
inputs = [0 for i in range(input_number)]

for i in range(input_number):
    inputs[i] = int(input())

# Finds responses
for i in inputs:
    doubled = i*2
    lower_bound = 2
    upper_bound = doubled-lower_bound
    found = False

    while not(found) and lower_bound <= i:
        if isPrime(lower_bound) and isPrime(upper_bound):
            print(str(lower_bound) + " " + str(upper_bound)) # prints answer if both bounds are prime numbers
            found = True

        if lower_bound == 2:
            lower_bound = lower_bound + 1
            upper_bound = upper_bound - 1
        else:
            lower_bound = lower_bound + 2
            upper_bound = upper_bound - 2


